### ISSUE TITLE
Invalid NTP protocol version number causing time sync to fail

### ISSUE THREAD

**Issue by bgh-github** [authelia/authelia#4772]
### Version

v4.37.5

### Deployment Method

Other

### Reverse Proxy

NGINX

### Reverse Proxy Version

N/A

### Description

When configured to use NTP version 4, Authelia writes a warning to the journal/log file when attempting to sync with certain NTP servers: `Could not read from the NTP server socket to validate the system time is properly synchronized: read udp <authelia ip>:<port>-><ntp server ip>:123: i/o timeout`

### Reproduction

This only seems to occur when the ntp>version key in the configuration is explicitly set to 4 or undefined (defaulting to 4). Switching the ntp>version value to 3 has been found to work around the problem.

Using `tcpdump` and `tshark` to observe client packets received by an internal `chronyd` NTP server upon restarting Authelia. Note unexpected reference to NTPv5

When config ntp>version == 4

```shell
> tcpdump src <authelia ip> and port ntp

... IP <authelia ip>.<port> > time.lan.example.com.ntp: NTPv5, Client, length 48

> tshark -Y "ip.src == <authelia ip>" -T pdml

<proto name="ntp" showname="Network Time Protocol (reserved, client)" size="48" pos="42">
  <field name="ntp.flags" showname="Flags: 0x2b, Leap Indicator: no warning, Version number: reserved, Mode: client" size="1" pos="42" show="0x2b" value="2b">
    <field name="ntp.flags.li" showname="00.. .... = Leap Indicator: no warning (0)" size="1" pos="42" show="0" value="0" unmaskedvalue="2b"/>
    <field name="ntp.flags.vn" showname="..10 1... = Version number: reserved (5)" size="1" pos="42" show="5" value="5" unmaskedvalue="2b"/>
    <field name="ntp.flags.mode" showname=".... .011 = Mode: client (3)" size="1" pos="42" show="3" value="3" unmaskedvalue="2b"/>
  </field>
```

In comparison to config ntp>version == 3

```shell
> tcpdump src <authelia ip> and port ntp

... IP <authelia ip>.<port> > time.lan.example.com.ntp: NTPv3, Client, length 48

> tshark -Y "ip.src == <authelia ip>" -T pdml

<proto name="ntp" showname="Network Time Protocol (NTP Version 3, client)" size="48" pos="42">
  <field name="ntp.flags" showname="Flags: 0x1b, Leap Indicator: no warning, Version number: NTP Version 3, Mode: client" size="1" pos="42" show="0x1b" value="1b">
    <field name="ntp.flags.li" showname="00.. .... = Leap Indicator: no warning (0)" size="1" pos="42" show="0" value="0" unmaskedvalue="1b"/>
    <field name="ntp.flags.vn" showname="..01 1... = Version number: NTP Version 3 (3)" size="1" pos="42" show="3" value="3" unmaskedvalue="1b"/>
    <field name="ntp.flags.mode" showname=".... .011 = Mode: client (3)" size="1" pos="42" show="3" value="3" unmaskedvalue="1b"/>
  </field>
```

The issue is also reprodicible when using certain other public NTP servers, for example
* clock.sjc.he.net
* time.windows.com
* time.nrc.ca

I'm not really familiar enough with the NTP packet structure to understand what may need to be changed.

### Expectations

When Authelia NTP is configured to use version 4, time sync against compliant NTP servers should succeed without warnings on startup.

### Configuration (Authelia)

```yaml
ntp:
  address: "time.lan.example.com:123"
  version: 4
```

### Logs (Authelia)

```shell
time="..." level=debug msg="Registering client <name> with policy two_factor (two_factor)"
time="..." level=debug msg="Registering client <name> with policy two_factor (two_factor)"
time="..." level=info msg="Storage schema is being checked for updates"
time="..." level=info msg="Storage schema is already up to date"
time="..." level=debug msg="notification provider: startup check skipped as it is disabled"
time="..." level=warning msg="Could not read from the NTP server socket to validate the system time is properly synchronized: read udp <authelia ip>:<port>-><ntp server ip>:123: i/o timeout"
time="..." level=info msg="Initializing server for TLS connections on '[::]:9091' path '/'"
```

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)
- [X] This is a bug report and not a support request
- [X] I have read the security policy and this bug report is not a security issue or security related issue
- [X] I have either included the complete configuration file or I am sure it's unrelated to the configuration
- [X] I have provided all of the required information in full with the only alteration being reasonable sanitization in accordance with the [Troubleshooting Sanitization](https://www.authelia.com/r/sanitize) reference guide
- [X] I have checked for related proxy or application logs and included them if available
- [X] I have checked for related issues and checked the documentation

**Comment by james-d-elliott** [comment]
Yeah I can see how this occurs, it's mostly because of the leap, version, and mode sharing a byte, and the method we used  to calculate this value. I will aim to produce a theoretical fix today.

**Comment by james-d-elliott** [comment]
Thanks for the detailed report by the way.

**Comment by james-d-elliott** [comment]
Give the `fix-ntp-version` tag  a shot. Alternatively the above PR has binaries in the [CI pipeline](https://buildkite.com/authelia/authelia/builds/18041).

**Comment by james-d-elliott** [comment]
This is because they're part of the same byte. Bits 0 and 1 align with Ll, bits 2, 3, and 4 align with the Vn, and bits 5, 6, 7 align with Mode. It's a very common pattern for languages to store them in a single byte (uint8) for this reason, as you can just use binary math to adjust the values.

When initially implemented I believe I hacked together a solution from memory of the spec since at the time there were no decent solutions for this, and I did this as an example for a person wanting to contribute, I am fairly sure I validated this I know I did check the code, pretty sure against chronyd actually (and have done so since). Interestingly it worked at one point with chronyd so maybe they were ignoring it for some time. I do find it interesting they have no errors on their end.

I clearly calculated the decimal values incorrectly in my head, and it's a clear example of why this should be done with binary / bitwise `AND` and 'OR' to avoid this kind of issue.

**Comment by james-d-elliott** [comment]
Thanks for letting me know, my testing returned the same results for the fix. I'll merge this now. As per several other issues we're in the middle of a feature release cycle and it's a large one, so this wont be released until 4.38, but I can backport the changes to a 4.37.5 branch if anyone is eager.
